### PR TITLE
Copter: Guided Mavlink Target Input loosen force_set restriction only for acceleration commands

### PR DESCRIPTION
**Pull request by hendjoshsr71**
#18720 was a bit too  restrictive considering how much pre-existing code is out there utilizing Force_Set `True` when acceleration targets aren't being commanded.   See this comment https://github.com/ArduPilot/ardupilot/issues/19462#issuecomment-989863780

This only rejects commands for when force_set is true & accelerations are to be commanded. 

Tested in SITL with:
`mode guided`
`module load message`
`arm throttle`
`takeoff 10`
`message SET_POSITION_TARGET_LOCAL_NED 0 1 0 1 4039 0 0 0 2 -2 -1 0 0 0 0 0`

Note above TYPE_MASK: sets ignore POS, ACCEL, Yaw, & Yaw Rate.  Sets Force_Set TRUE and allows Velocity commands

### CODE REVIEW

**Comment by lthall** [comment]
Sorry, why are we doing this. We do not support Force Set under any circumstances.

**Comment by hendjoshsr71** [comment]
So basically, my reasoning is force_set only effects acceleration inputs and the number of support issues has reached a number where I think this is good idea before making it into stable...

The reason for the support issues is everyone has naturally set FORCE_SET TRUE in all of their scripts where they are not using acceleration commands. (Because a type mask of all ones makes sense right, 111111111000)  See this for an example support https://github.com/ArduPilot/ardupilot/issues/19462#issuecomment-989863780

Examples: ROS scripts, example codes, Mission Planner (I fixed that once already)

Basically, we get stuck with extra code because the mavlink message reversed this field compared to every other field. MEH 

I'm ok if we don't do this but I foresee lots of support requests, "Of this worked on 4.1 stable but not on 4.2 stable..."

**Comment by lthall** [comment]
Yes, I too saw this.

I would suggest we remove the check for that bit all together instead.

If we check it then we must reject it. If we ignore it then we don't check it.

What do you think?

**Comment by hendjoshsr71** [comment]
I disagree. 

**"If"** someone reads the MAVLink spec and decides to command accelerations using units Newtons that could be a bad day for them.  

So either we need to handle the force_set bit or reject the command. That is a safety thing so in my mind that must stay in some form.

But we have a competing interest of old scripts from before we accepted acceleration commands that blindly set `force_set`. So that is why I'm proposing only rejecting commands for the case where it can cause safety problems. 

This bit only effects how we interpret acceleration commands thus this change.

Maybe discuss on Disord? Text isn't the best...

**Comment by lthall** [comment]
> "If" someone reads the MAVLink spec and decides to command accelerations using units Newtons that could be a bad day for them.

This change allows this precisely. Even worse, this code encourages a user that has checked the code quickly to see if we support it.

As I said early on when I started updating all the guided commands to properly support the mavlink commands, I am very strongly of the opinion that we support each message rigorously to specification. In cases where we have made a mistake we fix our mistake and expect the users to fix it in their code.

In this case our current implementation interoperates the message correctly and fully. I am willing to concede that we could ignore a bit completly if it doesn't make sense in our application but that is not my preference. 

What I am very strongly opposed to is bastardising messages because it makes things easy for lazy or sloppy users.

> But we have a competing interest of old scripts from before we accepted acceleration commands that blindly set force_set. 

Then they will fix the bug in their code. It is not sensible to fix this problem by introducing a bug into our code.

**Comment by lthall** [comment]
I went over this again and I am rolling back my position a little.

What this change does is only rejects commands for when force_set is true & **Force is commanded**.

This is fine. It meets spec while being maximally flexible in the interpretation of the command.

I think we just need to clarify the comment and we are good!

Sorry I did not fully apricate what you were doing.

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -1160,8 +1160,8 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         bool yaw_rate_ignore = packet.type_mask & MAVLINK_SET_POS_TYPE_MASK_YAW_RATE_IGNORE;
         bool force_set       = packet.type_mask & MAVLINK_SET_POS_TYPE_MASK_FORCE_SET;
 
-        // force_set true, do not accept command
-        if (force_set) {
+        // force_set is true when acceleration is to be commanded, do not accept command
```
Comment (lthall) [inline ArduCopter/GCS_Mavlink.cpp:1163]: I think we need to be a little clearer with this comment. Something like:

// Force inputs are not supported
// Do not accpt command if force_set is true and acc_ignore is false

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -1254,8 +1254,8 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
...
```
Comment (lthall) [inline ArduCopter/GCS_Mavlink.cpp:1257]: Same as above

**Review by lthall — CHANGES_REQUESTED** [review]
These changes open the possibility of a user providing a command that we do not support.

File: ArduCopter/GCS_Mavlink.cpp (hunk 1 above)
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp]: Definitely. Fixed.

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (hendjoshsr71) [reply ArduCopter/GCS_Mavlink.cpp]: Fixed

**Comment by hendjoshsr71** [comment]
> I went over this again and I am rolling back my position a little.
> 
> What this change does is only rejects commands for when force_set is true & **Force is commanded**.
> 
> This is fine. It meets spec while being maximally flexible in the interpretation of the command.
> 
> I think we just need to clarify the comment and we are good!
> 
> Sorry I did not fully apricate what you were doing.

No, problem. Sometimes I over-explain things which muddies the water a bit.

**Review by peterbarker — APPROVED** [review]

**Review by rmackay9 — APPROVED** [review]
